File: core/src/protocol.ts
/**
 * Menu bar shell ↔ helper protocol: newline-delimited JSON over the helper's
 * stdin/stdout. The Swift side mirrors these shapes in `Protocol.swift`; the
 * contract test (app/helper/test/protocol.test.ts) writes a fixture the Swift
 * test decodes, so drift fails CI on both sides.
 */
import type { RangeKey } from "./range.ts";
import type { Source } from "./types.ts";

export type SortKey = "tokens" | "parallelism" | "cost";

export interface View {
  range: RangeKey;
  groupId: number | null;
  sort: SortKey;
}

export type Command =
  | {
      id: number;
      cmd: "init";
      /** From the Keychain; sent over stdin so it never shows up in `ps`. */
      token: string | null;
      serverUrl: string;
      view: View;
      enabledSources: Source[];
    }
  | { id: number; cmd: "signUp"; name: string }
  | { id: number; cmd: "rename"; name: string }
  | { id: number; cmd: "createGroup"; name: string }
  | { id: number; cmd: "joinGroup"; code: string }
  | { id: number; cmd: "leaveGroup"; groupId: number }
  | { id: number; cmd: "rotateCode"; groupId: number }
  | { id: number; cmd: "setView"; view: View }
  | { id: number; cmd: "setSources"; enabledSources: Source[] }
  | { id: number; cmd: "syncNow" }
  | { id: number; cmd: "refresh" }
  /** Replies with `{ url }`: the dashboard, signed in via a single-use code. */
  | { id: number; cmd: "openDashboard" }
  | { id: number; cmd: "signOut" }
  /** Starts `brew upgrade` (the app quits and relaunches), or replies `{ url }` to download manually. */
  | { id: number; cmd: "installUpdate" };

export type Message =
  | { id: number; ok: true; result: unknown }
  /** Stable codes: name_taken, invalid_name, not_found, group_full, rate_limited, offline, … */
  | { id: number; ok: false; error: string }
  | { event: "state"; state: AppState }
  /** After signUp → save to Keychain. `null` → the token was rejected; delete it. */
  | { event: "token"; token: string | null };

export interface MeStats {
  name: string;
  /** Rank in the current view; null before the first leaderboard load. */
  rank: number | null;
  of: number;
  tokens: number;
  costUsd: number;
  parallelism: number | null;
  peakAgents: number;
  tokensPerActiveHour: number | null;
}

export interface LeaderboardRow {
  rank: number;
  name: string;
  tokens: number;
  costUsd: number;
  parallelism: number | null;
  peakAgents: number;
  isMe: boolean;
}

export interface GroupInfo {
  id: number;
  name: string;
  /** Display form, e.g. `K7QM-2XRP-9D`. */
  code: string;
  memberCount: number;
  isOwner: boolean;
}

export interface SourceInfo {
  id: Source;
  label: string;
  enabled: boolean;
}

export interface AppState {
  phase: "starting" | "onboarding" | "ready";
  version: string;
  view: View;
  me: MeStats | null;
  /** Top of the leaderboard; the shell pins `me` below it when outside. */
  leaderboard: LeaderboardRow[];
  groups: GroupInfo[];
  sync: {
    syncing: boolean;
    lastSyncedAt: number | null;
    lastError: string | null;
    online: boolean;
  };
  sources: SourceInfo[];
  /** A newer published version, or null. */
  update: {
    version: string;
    /** Installed with Homebrew, so the helper can upgrade in place. */
    viaBrew: boolean;
    installing: boolean;
  } | null;
}

export const SOURCE_LABELS: Record<Source, string> = {
  claude_code: "Claude Code",
  claude_cowork: "Claude Cowork",
  codex: "Codex",
  cursor_local: "Cursor",
};

export const LEADERBOARD_TOP = 10;

File: app/macos/Sources/Tokenmaxxing/Protocol.swift
import Foundation

// Mirrors core/src/protocol.ts. Tests/TokenmaxxingTests/Fixtures/messages.ndjson
// is written by the TypeScript side and decoded in ProtocolTests, so drift fails CI.

enum RangeKey: String, Codable, CaseIterable, Sendable {
    case today
    case week = "7d"
    case month = "30d"
    case all

    var label: String {
        switch self {
        case .today: "Today"
        case .week: "7d"
        case .month: "30d"
        case .all: "All"
        }
    }
}

enum SortKey: String, Codable, CaseIterable, Sendable {
    case tokens, parallelism, cost, prs

    var label: String {
        switch self {
        case .tokens: "Tokens"
        case .parallelism: "Parallel"
        case .cost: "Cost"
        case .prs: "PRs"
        }
    }
}

struct ViewSettings: Codable, Equatable, Sendable {
    var range: RangeKey
    var groupId: Int?
    var sort: SortKey

    static let `default` = ViewSettings(range: .today, groupId: nil, sort: .tokens)

    /// `groupId` must go out as `null`, not be omitted: the helper compares against null.
    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(range, forKey: .range)
        try c.encode(groupId, forKey: .groupId)
        try c.encode(sort, forKey: .sort)
    }
}

struct MeStats: Codable, Equatable, Sendable {
    let name: String
    let rank: Int?
    let of: Int
    let tokens: Double
    let costUsd: Double
    let parallelism: Double?
    let peakAgents: Int
    let tokensPerActiveHour: Double?
    let prs: Int
}

struct LeaderboardRow: Codable, Equatable, Identifiable, Sendable {
    let rank: Int
    let name: String
    let tokens: Double
    let costUsd: Double
    let parallelism: Double?
    let peakAgents: Int
    let prs: Int
    let isMe: Bool
    var id: String { name }
}

struct GroupInfo: Codable, Equatable, Identifiable, Sendable {
    let id: Int
    let name: String
    let code: String
    let memberCount: Int
    let isOwner: Bool
}

struct SourceInfo: Codable, Equatable, Identifiable, Sendable {
    let id: String
    let label: String
    let enabled: Bool
}

struct SyncInfo: Codable, Equatable, Sendable {
    let syncing: Bool
    let lastSyncedAt: Double?
    let lastError: String?
    let online: Bool
}

struct UpdateInfo: Codable, Equatable, Sendable {
    let version: String
    let viaBrew: Bool
    let installing: Bool
}

enum Phase: String, Codable, Sendable {
    case starting, onboarding, ready
}

struct AppState: Codable, Equatable, Sendable {
    let phase: Phase
    let version: String
    let view: ViewSettings
    let me: MeStats?
    let leaderboard: [LeaderboardRow]
    let groups: [GroupInfo]
    let sync: SyncInfo
    let sources: [SourceInfo]
    let update: UpdateInfo?
}

/// The few fields any command result carries (`openDashboard` → url, group actions → code).
struct CommandResult: Decodable, Sendable {
    let url: String?
    let code: String?
    let name: String?
}

/// One line from the helper: a reply (`id`) or an event (`event`).
struct IncomingMessage: Decodable, Sendable {
    let id: Int?
    let ok: Bool?
    let error: String?
    let result: CommandResult?
    let event: String?
    let state: AppState?
    let token: String?
}

/// One line to the helper. Nil fields are omitted from the JSON.
struct OutgoingCommand: Encodable, Sendable {
    var id = 0
    let cmd: String
    var token: String? = nil
    var serverUrl: String? = nil
    var view: ViewSettings? = nil
    var enabledSources: [String]? = nil
    var name: String? = nil
    var code: String? = nil
    var groupId: Int? = nil
}
